caffe2/operators/batch_bucketize_op.cc

```diff
@@ -0,0 +1,126 @@
+#include "batch_bucketize_op.h"
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/tensor.h"
+
+namespace caffe2 {
+
+template <>
+bool BatchBucketizeOp<CPUContext>::RunOnDevice() {
+  auto& indices = Input(INDICES);
+  auto& borders = Input(BORDERS);
+  auto& lengths = Input(LENGTHS);
+  auto& feature = Input(FEATURE);
+  auto* output = Output(O);
+  CAFFE_ENFORCE_EQ(lengths.ndim(), 1);
+  CAFFE_ENFORCE_EQ(indices.ndim(), 1);
+  CAFFE_ENFORCE_EQ(borders.ndim(), 1);
+  CAFFE_ENFORCE_EQ(feature.ndim(), 2);
+  CAFFE_ENFORCE_EQ(lengths.size(), indices.size());
+
+  const auto* lengths_data = lengths.template data<int32_t>();
+  const auto* indices_data = indices.template data<int32_t>();
+  const auto* borders_data = borders.template data<float>();
+  const auto* feature_data = feature.template data<float>();
+  auto batch_size = feature.dim(0);
+  auto feature_dim = feature.dim(1);
+  auto output_dim = indices.size();
+
+  TIndex length_sum = 0;
+  for (TIndex i = 0; i < lengths.size(); i++) {
+    CAFFE_ENFORCE_GE(feature_dim, indices_data[i]);
+    length_sum += lengths_data[i];
+  }
+  CAFFE_ENFORCE_EQ(length_sum, borders.size());
+
+  TIndex lower_bound = 0;
+  output->Resize(batch_size, output_dim);
+  auto* output_data = output->template mutable_data<int32_t>();
+
+  for (TIndex i = 0; i < batch_size; i++) {
+    lower_bound = 0;
+    for (TIndex j = 0; j < output_dim; j++) {
+      for (TIndex k = 0; k <= lengths_data[j]; k++) {
+        if (k == lengths_data[j] ||
+            feature_data[i * feature_dim + indices_data[j]] <
+                borders_data[lower_bound + k]) {
+          output_data[i * output_dim + j] = k;
+          break;
+        } else {
+          continue;
+        }
+      }
+      lower_bound += lengths_data[j];
+    }
+  }
+  return true;
+}
+
+REGISTER_CPU_OPERATOR(BatchBucketize, BatchBucketizeOp<CPUContext>);
+
+OPERATOR_SCHEMA(BatchBucketize)
+    .NumInputs(4)
+    .NumOutputs(1)
+    .SetDoc(R"DOC(
+Bucketize the float_features into sparse features.
+The float_features is a N * D tensor where N is the batch_size, and D is the feature_dim.
+The indices is a 1D tensor containing the indices of the features that need to be bucketized.
+The lengths is a 1D tensor that splits the following 'borders' argument.
+The borders is a 1D tensor containing the border list for each feature.
+
+With in each batch, `indices` should not have duplicate number,
+and the number of elements in `indices` should be less than or euqal to `D`.
+Each element in `lengths` vector (lengths[`i`]) represents
+the number of borders in the sub border list.
+The sum of all elements in `lengths` must be equal to the size of  `borders`.
+If lengths[0] = 2, the first sub border list is [0.5, 1.0], which separate the
+value to (-inf, 0.5), [0,5, 1.0), [1.0, inf). The bucketized feature will have
```

Comment:
Could you do right-inclusive like in `BatchBucketOneHot`

Reply:
Will change that.